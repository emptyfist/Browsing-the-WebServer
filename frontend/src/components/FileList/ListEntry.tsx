import React from 'react';
import { useLocation, useNavigate } from 'react-router-dom';
import { FileInfo } from '../../types';
import FolderIcon from '../../assets/svg/folder.svg';
import FileIcon from '../../assets/svg/file.svg';
import TrashIcon from '../../assets/svg/trash.svg';

interface Props {
  data: FileInfo,
  refresh: (path: string) => void,
  showConfirm: (name: string) => void
}

const BE_URL = process.env.BACKEND_URL ?? 'http://localhost:8088';

const ListEntry = ({data, refresh, showConfirm}: Props) => {
  const navigate = useNavigate();
  const location = useLocation();

  const decorateFileSize = (val: number) => {
    if (val < 1024) return `${val}byte`;
    else if (val < 1024 * 1024) return `${Math.ceil(val / 1024)}KB`;
    else if (val < 1024 * 1024 * 1024) return `${Math.ceil(val / 1024 / 1024)}MB`;
    else if (val < 1024 * 1024 * 1024 * 1024) return `${Math.ceil(val / 1024 / 1024 / 1024)}GB`;
    else return `${Math.ceil(val / 1024 / 1024 / 1024 / 1024)}TB`;
  }

  return (
    <div className="list-entry-wrapper">
      <div className="list-entry">
        {data.Type === 'File' ? <img src={FileIcon} alt="folder" /> : <img src={FolderIcon} alt="file" />}
        {data.Type === 'File' ? 
          <a 
            className='entry-name' 
            href={`${BE_URL}/api/file/download?path=${location.pathname === '/' ? '' : location.pathname}/${data.Name}`} 
            download
          >
            {`${data.Name}   ${decorateFileSize(data.Size)}`}
          </a> : 
          <span 
            className='entry-name' 
            onClick={e=> {
              const path = `${location.pathname === '/' ? '' : location.pathname}/${data.Name}`;
              navigate(path);
              if (refresh) refresh(path);
            }}
          >
            {data.Name}
          </span>
        }
      </div>
      <button 
        className="delete-btn" 
        // disabled={data.Type !== 'File'}
        onClick={(e)=>{
          if (showConfirm) showConfirm(data.Name);
        }}
      >
        <img src={TrashIcon} alt="Delete" />
      </button>
    </div>
  )
}

export default ListEntry;
