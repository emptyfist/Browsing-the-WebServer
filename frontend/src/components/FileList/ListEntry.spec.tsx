import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';

import ListEntry from './ListEntry';

describe('ListEntry test', function () {
  it('should return Directory SnapShot', function () {
    const folderEntry = {
      Type: 'Directory',
      Name: 'SubDirectory',
      Size: 0
    };

    const { container } = render(
      <MemoryRouter>
        <ListEntry data={folderEntry} refresh={() => null} showConfirm={() => null}/>
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });

  it('should return FileEntry SnapShot', function () {
    const fileEntry = {
      Type: 'File',
      Name: 'TestFile',
      Size: 10254
    };

    const { container } = render(
      <MemoryRouter>
        <ListEntry data={fileEntry} refresh={() => null} showConfirm={() => null}/>
      </MemoryRouter>
    );

    expect(container).toMatchSnapshot();
  });
});
