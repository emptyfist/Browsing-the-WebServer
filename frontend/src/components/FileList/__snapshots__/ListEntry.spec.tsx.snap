// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListEntry test should return Directory SnapShot 1`] = `
<div>
  <div
    class="list-entry-wrapper"
  >
    <div
      class="list-entry"
    >
      <img
        alt="file"
        src="folder.svg"
      />
      <span
        class="entry-name"
      >
        SubDirectory
      </span>
    </div>
    <button
      class="delete-btn"
    >
      <img
        alt="Delete"
        src="trash.svg"
      />
    </button>
  </div>
</div>
`;

exports[`ListEntry test should return FileEntry SnapShot 1`] = `
<div>
  <div
    class="list-entry-wrapper"
  >
    <div
      class="list-entry"
    >
      <img
        alt="folder"
        src="file.svg"
      />
      <a
        class="entry-name"
        download=""
        href="http://localhost:8080/api/file/download?path=/TestFile"
      >
        TestFile   11KB
      </a>
    </div>
    <button
      class="delete-btn"
    >
      <img
        alt="Delete"
        src="trash.svg"
      />
    </button>
  </div>
</div>
`;
